overlord/state/state.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2022 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/juju/errors"
+)
+
+type State struct {
+	db *sql.DB
+}
+
+func NewState(db *sql.DB) *State {
+	return &State{
+		db: db,
+	}
+}
+
+// DB returns the underlying sql.DB.
+// It is expected to use the transaction of state, but sometimes a raw database
+// is required in some scenarios.
+func (s *State) DB() *sql.DB {
+	return s.db
+}
+
+// PrepareStatement creates a prepared statement for later queries or
+// executions.
+// Multiple queries or executions may be run concurrently from the
+// returned statement.
+// The caller must call the statement's Close method
+// when the statement is no longer needed.
+func (s *State) PrepareStatement(ctx context.Context, query string) (*sql.Stmt, error) {
+	return s.db.PrepareContext(ctx, query)
+}
+
+type Txn interface {
+	// PrepareContext creates a prepared statement for use within a transaction.
+	//
+	// The returned statement operates within the transaction and will be closed
+	// when the transaction has been committed or rolled back.
+	//
+	// To use an existing prepared statement on this transaction, see Tx.Stmt.
+	//
+	// The provided context will be used for the preparation of the context, not
+	// for the execution of the returned statement. The returned statement
+	// will run in the transaction context.
+	PrepareContext(context.Context, string) (*sql.Stmt, error)
+	// StmtContext returns a transaction-specific prepared statement from
+	// an existing statement.
+	//
+	// Example:
+	//  updateMoney, err := db.Prepare("UPDATE balance SET money=money+? WHERE id=?")
+	//  ...
+	//  tx, err := db.Begin()
+	//  ...
+	//  res, err := tx.StmtContext(ctx, updateMoney).Exec(123.45, 98293203)
+	//
+	// The provided context is used for the preparation of the statement, not for the
+	// execution of the statement.
+	//
+	// The returned statement operates within the transaction and will be closed
+	// when the transaction has been committed or rolled back.
+	StmtContext(context.Context, *sql.Stmt) *sql.Stmt
+	// ExecContext executes a query that doesn't return rows.
+	// For example: an INSERT and UPDATE.
+	ExecContext(context.Context, string, ...interface{}) (sql.Result, error)
+	// QueryContext executes a query that returns rows, typically a SELECT.
+	QueryContext(context.Context, string, ...interface{}) (*sql.Rows, error)
+}
+
+// TxnBuilder allows the building of a series of functions that will be called
+// during a transaction commit. Only upon commit is the transaction constructed
+// and the function called.
+// The functions in the txn builder maybe called multiple times depending on
+// how many retries are employed.
+type TxnBuilder interface {
+	Run(func(context.Context, Txn) error)
```

Comment:
Why not make this return `TxnBuilder` so we can chain these?

Reply:
Now that I think about this, maybe we should call this `Queue`, `QueueStep` or `Stage` to avoid confusing folks given that the manager has a `Run` method that auto-commits?

Reply:
Yeah, I was struggling with this naming.